Framework/PythonInterface/plugins/algorithms/FitIncidentSpectrum.py

```diff
@@ -67,36 +72,37 @@ def PyInit(self):
     def _setup(self):
         self._input_ws = self.getProperty('InputWorkspace').value
         self._output_ws = self.getProperty('OutputWorkspace').valueAsStr
+        self._incident_index = self.getProperty('WorkspaceIndex').value
+        self._binning_for_calc = self.getProperty('BinningForCalc').value
         self._binning_for_fit = self.getProperty('BinningForFit').value
         self._fit_spectrum_with = self.getProperty('FitSpectrumWith').value
-        self._binning_for_calc = self.getProperty('BinningForCalc').value
-        if not self._binning_for_calc.all():
-            x = self._input_ws.readX(0)
-            self._binning_for_calc = [str(i) for i in [min(x), x[1] - x[0], max(x) + x[1] - x[0]]]
 
     def PyExec(self):
         self._setup()
-
-        x = np.arange(self._binning_for_calc[0], self._binning_for_calc[2], self._binning_for_calc[1])
-        incident_index = 0
+        if self._binning_for_calc.size == 0:
+            x = np.array(self._input_ws.readX(self._incident_index))
+            self._binning_for_calc = [i for i in [min(x), x[1] - x[0], max(x) + x[1] - x[0]]]
+        else:
+            x = np.arange(self._binning_for_calc[0], self._binning_for_calc[2], self._binning_for_calc[1])
         if self._binning_for_fit.size == 0:
-            x_fit = np.array(self._input_ws.readX(incident_index))
-            y_fit = np.array(self._input_ws.readY(incident_index))
+            x_fit = np.array(self._input_ws.readX(self._incident_index))
+            y_fit = np.array(self._input_ws.readY(self._incident_index))
         else:
             rebinned = Rebin(
                 self._input_ws,
                 Params=self._binning_for_fit,
                 PreserveEvents=True,
                 StoreInADS=False)
-            x_fit = np.array(rebinned.readX(incident_index))
-            y_fit = np.array(rebinned.readY(incident_index))
+            x_fit = np.array(rebinned.readX(self._incident_index))
+            y_fit = np.array(rebinned.readY(self._incident_index))
 
+        x_bin_centers = [(x[i+1]+x[i])/2.0 for i in range(x.size-1)]
```

Comment:
I think you might be able to do this quicker with numpy array arithmetic rather than a list comprehension?
```python
0.5*(x[:-1] + x[1:])
```